Add environment / configuration mocking for flexible configuration and controlled testing

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Supersedes #661, and only focuses on adding the mock environment while removing the refactored code.  As a result, all additions are only focused on adding mock and tests to the existing environment, and are not as worried about modifying, adjusting, or otherwise correcting any underlying implementation.

These Mocks are meant to provide primitive mock implementations of various systems with the goal of simplifying the interactions between these various systems down to idealized cases.  The mock environment can be created without any configuration being passed, yet also has the ability for specific configuration parameters to be overwritten using the [options pattern](https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html).  This should allow for good sensible defaults, with specifics able to be overwritten should the need arise.

Every Mock feature addition features comments, and unit tests or Examples that demonstrate their construction or intended usage.

This PR aims to be purely additive, and takes into consideration much of the feedback from the previous PR.  It is created separately so as to avoid confusion.

### This PR does not:
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->
Make any modification to the implemented Espresso solution within the `TransactionStreamer`.

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->
<!-- [ ] For integration projects, make sure it won't break backward compatibility. -->

<!-- Details on maintaining backward compatibility for integration projects: -->
<!-- * Try to keep changes additive: add new, optional methods, flags, or parameters instead of modifying or removing existing functionality. -->
<!-- * If modification is necessary, it should be either a clear bug fix or guarded by a config/feature flag.  -->
<!-- * Follow Open Closed Principle and Interface Segregation Principle, clients should not be forced to depend on interfaces they do not use. -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->

Ayiga committed:
Add UnimplementedEspressoClient

In order to add flexibility to our testing approach, we need to control certain elements
of our environment.  One of these elements is the EspressoClient itself.  The
EspressoClient serves as an interface that allows the consumer to interact with the
Espresso Chain.  This is usually done via an API request, but it doesn't necessarily
need to mean that.

Certain aspects of the EspressoClient's behavior could potentially have a profound
impact our implementation and as such we need to be able to control some aspects
about the EspressoClient, without necessarily hitting a live Espresso Chain,  or
worrying about all of the other aspects of the EspressoClient.

As a result, and to make it easier to implement some future implementations the
UnimplementedEspressoClient has been defined in order to make it easier to
write Mocks of the EspressoClient without the need to fulfill the full interface.

Ayiga committed:
Add UnimplementedLightClientReader

In order to add flexibility to our testing approach, we need to control certain elements
of our environment.  One of these elements is the LightClientReaderInterface.  The
LightClientReaderInterface represents a type that is interacting with a smart contract
on an RPC endpoint, and reading it's relevant state.

Certain aspects of the LightClientReaderInterface will affect how our logic progresses
and it is beneficial to be able to have direct control over the results this interface
returns.

As a result, and to make it easier to implement some future implementations the
UnimplementedLightClientReader has been defined in order to make it easier to
write Mocks of the LightClientReaderInterface without the need to fulfill the full
interface.

Ayiga committed:
Add UnimplementedExecutionClient

In order to add flexibility to our testing approach, we need to control certain elements
of our environment.  One of these elements is the ExecutionClient.  The
ExecutionClient governs how information is recalled from, and flows through the
L2 in a predictable manner.

The ExecutionClient has a profound impact on our integration efforts as it is
directly responsible for message processing within the system.  In order to ensure
that our implementation makes sense, it is beneficial to be able to have a simplified
implementation of the ExecutionClient that behaves in a controllable / predicable
manner.

As a result, and to make it easier to implement some future implementations the
UnimplementedExecutionClient has been defined in order to make it easier to
write Mocks of the ExecutionClient without the need to fulfill the full
interface.

Ayiga committed:
Add MessageHasher

In order to simplify Message processing, it is beneficial to have a consistent way to
generate a Hash for a given `MessageWithMetadata` input.  This way when we write
interactions / tests, we need not worry about the specifics of how we achieve these
hashes, provided that we do it consistently.

This commit adds a `MessageHasher` that does just this, and provides two separate
implementations for the convenience of testing.  It is likely that the user shouldn't
need to worry about which of these to use, so in addition a `DefaultMessageHasher`
has been defined so that the user doesn't need to worry about the specifics.

In general, and in the future, whenever a `MessageHasher` is needed, we will
default to the `DefaultMessageHasher`, while allowing the user to swap the
implementation out, should they need to.

Ayiga committed:
Add MockExecutionEngineForTransactionStreamer

The `TransactionStreamer` utilizes an `ExecutionClient` in its operation, but it only
calls a subset of the `ExecutionClient` interface.  While the `ExecutionClient` plays
an important role, we can provide a simplified Mock implementation that only needs to
cater to the specifics, as understood by the `TransactionStreamer`.

The `MockExecutionEngineForTransactionStreamer` does just this by implementing
the minimum methods needed for the `TransactionStreamer` in an effort to provide
the implementation with the simplest implementation possible.  This **should** help
to eliminate any `ExecutionClient` implementation errors / details from consideration
with the Espresso specific interaction of the `TransactionStreamer`.

Ayiga committed:
Add MockEspressoKeyManager

The `EspressoKeyManagerInterface` plays an important role in the signing of
transaction data in order to authenticate and verify it when it comes back from
the Espresso Chain.  Creating the current implementation requires a minimum of
6 arguments, and it may be difficult to understand the how and why of each field
at a glance.  As a result, it is cumbersome to utilize and configure effectively.  Additionally, it provides not control over the private key that is utilized for signing
data, which makes testing with a result that is reproducible consistently difficult.

The `MockEspressoKeyManager` implements the `EspressoKeyManagerInterface`
interface, and allows some simple configuration options, with some sensible defaults
that can be overwritten by the user if desired.  This allows for a much easier to
setup `EspressoKeyManagerInterface`, that is only concerned with signing things
and does not worry about the other aspects.

Ayiga committed:
Add MockAlwaysLiveLightClientReader

The `TransactionStreamer` currently only ever utilizes the
`LightClientReaderInterface` in order to check liveness of `HotShot`.  In the majority
of our test cases, we want `HotShot` to be considered live. As a result, it is beneficial
to remove the complications of setting up Smart Contract reading that is dependent
on network and chain conditions, and to instead just utilize a replacement that just
always considers `HotShot` as live, with no delay in communication.

Ayiga committed:
Add MockEspressoChain

When testing our implementation that performs finalization checks against Espresso, it
is beneficial to operate on a controllable idealized version the Espresso Chain that
operates and documents our initial assumptions of how the chain should progress.
Additionally, it is beneficial to have direct control over this chain so that we are able
to control exactly when and how it advances, and the rate at which it produces blocks.

This simplified idealized representation of the `EspressoChain`, implements only the
methods needed for operation and utilization within the `TransactionStreamer`
interface.

Ayiga committed:
Add EspressoClientSimulatedLatency

When interacting with an idealized `EspressoClient`, it becomes prudent to test
different scenarios in terms of network performance, and their affect on our
implementations progression.  As a result, it is beneficial to be able to simulate
latency that is incurred in the process via various means, such as Network
calls and the like.

The `EspressoClientSimulatedLatency` allows for the configuration of delays against
the three primarily utilized endpoints of the `EspressoClient` for the
`TransactionStreamer` interaction.  It also includes a delay that can be configured
against the Explorer endpoint, that is utilized as a temporary fix for performance issues
discovered live.

Ayiga committed:
Add SiphonBlocksWithTransactions

When interacting with the Espresso Chain, it becomes beneficial to be able to inspect
what is being retrieved from the Espresso Chain itself.  This allows one the ability to
inspect the response, and check it against the logic that consumes in other aspects
such as the resubmission, and finality checking processes.

In order to facilitate such a need, the `SiphonBlocksWithTransactions` has been
written to provide this data so that the individual `TransactionsInBlock` responses
can be inspected, and checked for consistency.

This also has the added benefit of knowing immediately when this data is being
provided, and should be considered "available" for the system as a whole.

Ayiga committed:
Add MessageGenerator

In order to be able to test the system effectively in a simulated scenario, we need to
be able to easily submit messages into the system so that they can simulate load
in an easily controllable manner.

We add the ability to easily generate new messages sequentially to test against the
`TransactionStreamer`.  These messages, can be generated and submitted at
controlled rates that can be further expanded in the future.

Ayiga committed:
Add helper configuration functions for TransactionStreamer

Many of the fields within the `TransactionStreamer` that facilitate our Espresso
integration are a mix of public and private, which makes them difficult or impossible
to modify from an external package for testing purposes.

This file adds a simple way of configuring all of the configurable Espresso fields
without the need to make them specifically public.  It also helps to keep these
fields separate from the rest of the `TransactionStreamer`, so that the
`TransactionStreamer` specifics do not need to be exposed.

Ayiga committed:
Add NewMockTransactionStreamerEnvironment function

In order to make the Espresso incorporation with the `TransactionStreamer` easier to
test, it helps to be able to easily spin up different scenarios for environments that
provides sensible defaults which allow basic invocation to work, while still allowing
for the flexibility to overwrite specifics.

The `NewMockTransactionStreamerEnvironment` provides the ability to do this.
Calling it without any options will still result in an implementation that can effectively
simulate these interactions, and allows the user to just focus on the results or the
specific criteria they are choosing to focus on, without needing to worry about
every individual detail.

Ayiga committed:
Add TestTransactionStreamerEspressoThroughput

This test is setup to measure the implementation performance of the Espresso Chain
against the Transaction Streamer in the current integration.

This demonstrates how to use the mock setup to run the tests, and how we can
measure the performance using the utilities we have defined.  This also demonstrates
the configurable nature of the mock environment, and how one can utilize the
configuration to get meaningful measurements out of the system.

Ayiga committed:
Fix golangci-lint issues

There are a quite a few lint errors being reported by `golangci-lint` concerning error
checking, and overflow bounds issues.  I'm fairly confident about the overflow issues
being non-issues, but to appease the linter, I've switched up the iteration to ignore
prevent the overflow issue entirely.  The error checks, likewise, *should* be fine, but
in order to appease the linter, I've switched to using `errors.Is` instead of `==` as
requested.

Additionally, I've added unused variable assignment explicitly to avoid the lint error
concerning unchecked return types.

Sneh1999 commented on arbnode/espresso_transaction_streamer_helpers.go:
```diff
@@ -0,0 +1,173 @@
+
+// WithEspressoClient is a functional option to set the Espresso client in the
+// TransactionStreamerEspressoConfig.
+func WithEspressoClient(espressoClient espresso_client.EspressoClient) TransactionStreamerEspressoOption {
```
why do we need functions for EspressoClient, LightClientReader etc if they are all public variables and can be set directly?

Sneh1999 commented on system_tests/espresso/chain/mock_espresso_chain.go:
```diff
@@ -0,0 +1,353 @@
+	block, blockOk := m.BlockHeightStore[blockHeight]
+	m.Lock.RUnlock()
+	if !blockOk {
+		return espresso_client.TransactionsInBlock{}, ErrorBlockNotFoundForHeight{Height: blockHeight}
```
I like this way of representing errors!

Sneh1999 commented on system_tests/espresso/chain/simulated_latency.go:
```diff
@@ -0,0 +1,145 @@
+
+// WithTransactionsInBlockDelay sets the delay for the FetchTransactionsInBlock
+// method.
+func WithTransactionsInBlockDelay(delay time.Duration) SimulatedLatencyOption {
```
same here, why do we need setter functions for public variables?

Sneh1999 commented on system_tests/espresso/chain/simulated_latency.go:
```diff
@@ -0,0 +1,145 @@
+// with the specified chain client and delay duration. The delay will be applied
+// to all method calls made to the chain client.
+func NewEspressoClientSimulatedLatency(client espresso_client.EspressoClient, options ...SimulatedLatencyOption) *EspressoClientSimulatedLatency {
+	const defaultDelay = 250 * time.Millisecond
```
nit: is there a way to make this delay configurable for testing?

Sneh1999 reviewed (approved):
Thanks @Ayiga , overall looks great!

I was wondering if we should add a readme inside `espresso` folder to explain and add all the functionality of this test suite and how to run it?

Ayiga committed:
Add TimingData nd Timing function / constructor

In order to effectively time things, it helps to have a convenient and consistent way to
represent timing data. The timing data is meant to be as simple as possible, and is not
meant to be constructed directly, though there is nothing stopping the user from doing
just that.

Ayiga commented:
> I was wondering if we should add a readme inside `espresso` folder to explain and add all the functionality of this test suite and how to run it?

This is not a bad idea.  But I'll leave it as future TODO work.
